lib/Dialect/Arc/Transforms/LowerVerifSimulations.cpp

```diff
@@ -0,0 +1,205 @@
+  auto &funcBody = funcOp.getBody().emplaceBlock();
+  builder.setInsertionPointToEnd(&funcBody);
+
+  auto falseOp = builder.create<hw::ConstantOp>(op.getLoc(), i1Type, 0);
+  auto trueOp = builder.create<hw::ConstantOp>(op.getLoc(), i1Type, 1);
+  auto lowOp = builder.create<seq::ToClockOp>(op.getLoc(), falseOp);
+  auto highOp = builder.create<seq::ToClockOp>(op.getLoc(), trueOp);
+
+  // Instantiate the implementation module.
+  auto instType = SimModelInstanceType::get(
+      context, FlatSymbolRefAttr::get(implOp.getSymNameAttr()));
+  auto instOp = builder.create<SimInstantiateOp>(op.getLoc());
+  auto &instBody = instOp.getBody().emplaceBlock();
+  auto instArg = instBody.addArgument(instType, op.getLoc());
+  builder.setInsertionPointToEnd(&instBody);
+
+  // Create an `scf.execute_region` op inside such that we can use simple
+  // control flow in the `arc.sim.instantiate` op body. This is simpler than
+  // setting up an `scf.while` op.
+  auto execOp = builder.create<scf::ExecuteRegionOp>(op.getLoc(), TypeRange{});
+  builder.setInsertionPointToEnd(&execOp.getRegion().emplaceBlock());
+
+  // Apply the initial clock tick to the design.
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, clockName, lowOp);
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, initName, trueOp);
+  builder.create<SimStepOp>(op.getLoc(), instArg);
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, clockName, highOp);
+  builder.create<SimStepOp>(op.getLoc(), instArg);
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, clockName, lowOp);
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, initName, falseOp);
+  builder.create<SimStepOp>(op.getLoc(), instArg);
+
+  // Create the block that will perform a single clock tick.
+  auto &loopBlock = execOp.getRegion().emplaceBlock();
+  builder.create<cf::BranchOp>(op.getLoc(), &loopBlock);
+  builder.setInsertionPointToEnd(&loopBlock);
+
+  // Sample the done and exit code signals.
+  auto doneSample =
+      builder.create<SimGetPortOp>(op.getLoc(), i1Type, instArg, doneName);
+  auto exitCodeSample = builder.create<SimGetPortOp>(op.getLoc(), exitCodeType,
+                                                     instArg, exitCodeName);
+
+  // Apply a full clock cycle to the design.
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, clockName, highOp);
+  builder.create<SimStepOp>(op.getLoc(), instArg);
+  builder.create<SimSetInputOp>(op.getLoc(), instArg, clockName, lowOp);
+  builder.create<SimStepOp>(op.getLoc(), instArg);
+
+  // If done, exit the loop.
+  auto &exitBlock = execOp.getRegion().emplaceBlock();
+  builder.create<cf::CondBranchOp>(op.getLoc(), doneSample, &exitBlock,
+                                   &loopBlock);
+  builder.setInsertionPointToEnd(&exitBlock);
+
+  // Compute `exit_code | (exit_code != 0)` as a way of guaranteeing that
+  // the exit code will be non-zero even if bits are truncated by the operating
+  // system.
```

Comment:
Smart way to deal with the fact that exit typically ANDs with 0xFF. Maybe worth mentioning in the simulation op documentation that on non-zero exit code the least significant bit is always set? But I guess also not really, because we are very vague about that anyway.

Reply:
Yeah I'm less and less convinced that we should actually have a multi-bit exit code at all. Maybe a `success` line would be better…